torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,6 +1158,76 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   abort(abortReason);
 }
 
+void dump_debugging_info() {
+  LOG(ERROR) << dump_nccl_trace();
+}
+
+size_t hashHeartBeat(const HeartBeat& heartBeat) {
+  std::hash<size_t> id_hasher;
+  std::hash<std::uint8_t> type_hasher;
+  return id_hasher(heartBeat.pg_id_) ^ id_hasher(heartBeat.seq_id_) ^
+      type_hasher(static_cast<std::uint8_t>(heartBeat.op_type_));
+}
+
+void ProcessGroupNCCL::ncclCommsMonitor() {
```

Comment:
One easy way to write a test for the watcher is, in c++, acquire the workMetaListMutex_ lock from inside the test code, which would stall the workCleanupLoop.  Then the workCleanupLoop would forget to issue the heartbeat and the test code should expect an exception from the watcher thread

Reply:
ok, I see.. Yeah, we can start with that as a unit test.